### PR TITLE
Add skill history, file analysis storage, and top ranked project summaries

### PR DESCRIPTION
**Pull request by Puneet-Maan**
## 📝 Description
Adds the next set of enhancements to Project Insights.

### Summary

Introduces new functionality in `project_insights.py`:

- Storing detailed file analysis for each insight (file counts, sizes, types, largest/newest file)
- Providing a chronological skill history via `list_skill_history`
- Adding summaries for top ranked projects through `summaries_for_top_ranked_projects`
- Normalizing and sorting skills for consistent ordering (including older entries)
- Treating valid but incorrect JSON shapes as corrupted and stashing backups safely

Expands test suite **test_project_insights.py** with coverage for:

- Chronological skill history  
- File analysis fields  
- Summaries for top ranked projects (including score)  
- Non list JSON corruption handling  

Closes: #67 #112 #116

---

## 🔧 Type of Change

- [x] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [x] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

### Automated

- python3 -m unittest discover -s test -p "test_project_insights.py" -v
- All tests in `test_project_insights.py` now pass, covering:
  - Chronological skill history
  - File analysis persistence
  - Top ranked summaries and score
  - Stashing corrupted/non list JSON
  - All previously existing behaviors

### Manual
- Verified new fields appear correctly in the JSON log
- Confirmed ranking summaries reflect accurate scores and contributor stats
- Checked that corrupted or malformed JSON logs are safely backed up

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [x] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [x] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

image

### CODE REVIEW

File: src/project_insights.py [hunk 1]
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
...
+                stack.append(child)
+
+    return files
+
+
+def _safe_int(value: Any) -> int:
+    try:
+        return int(value)
+    except Exception:
+        return 0
```
Comment (Weebtrain) [inline src/project_insights.py]: No docstring explaining this method ensures int returned without error

File: src/project_insights.py [hunk 2]
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
...
+def _parse_timestamp(value: Any) -> Optional[datetime]:
+    if not isinstance(value, str):
+        return None
+    try:
+        return datetime.fromisoformat(value)
+    except ValueError:
+        try:
+            return datetime.strptime(value, "%Y-%m-%d %H:%M:%S")
+        except Exception:
+            return None
```
Comment (Weebtrain) [inline src/project_insights.py]: No doctring or comments explaining what this does. Good use of safe data conversion.

File: src/project_insights.py [hunk 3]
```diff
@@ -132,6 +133,102 @@ def _parse_analyzed_at(ts: str) -> datetime:
...
+            continue
+        node_type = str(node.get("type", "")).upper()
+        children = node.get("children") or []
+        if node_type and node_type != "DIR":
+            files.append(node)
+        for child in children:
+            if isinstance(child, dict):
+                stack.append(child)
+
+    return files
```
Comment (Weebtrain) [inline src/project_insights.py]: Useful method. Would serve better as a public method in a utility file for use in other places.

File: src/project_insights.py [hunk 4]
```diff
@@ -309,9 +414,61 @@ def rank_projects_by_contribution(
...
+    """
+    return [
+        {
+            "project_name": insight.project_name,
+            "skills": list(insight.skills),
+            "analyzed_at": insight.analyzed_at,
+            "skill_count": len(insight.skills),
+        }
+        for insight in list_project_insights(storage_path)
+    ]
```
Comment (Weebtrain) [inline src/project_insights.py]: Docstring is not consistent in listing the return structure, missing "skill_count" mention. No explenation  of parameters.

File: test/test_project_insights.py [hunk 5]
```diff
@@ -9,6 +9,7 @@
 
 import gc
 import json
+import logging
```
[the remark this answers is not in the document]
Comment (Puneet-Maan) [reply test/test_project_insights.py]: I think this is part of the python standard library

**Comment by Puneet-Maan** [comment]
@Weebtrain Hey! Thanks for the thorough review. I've pushed updates addressing all your points:
Documentation updates:
- Added full docstrings (with Args/Returns) to _safe_int, _parse_timestamp, _flatten_file_nodes, and _compute_file_analysis
- Expanded docstrings for list_skill_history and summaries_for_top_ranked_projects to list all parameters and returned fields
- Added clearer inline comments in _compute_file_analysis, especially around how the biggest and newest files are chosen
- Explained how storage_path is used inside summaries_for_top_ranked_projects

Test improvements:
- Added comments grouping related assertions and explaining what each group is checking
- Made edge case intentions clearer (like why Beta appears before Alpha in the time based tests)

File: src/project_insights.py [hunk 6]
```diff
@@ -277,7 +453,15 @@ def record_project_insight(
...
-    """Return all stored insights ordered chronologically."""
+    """
+    Return all stored insights in chronological order (oldest → newest).
+
+    Args:
+        storage_path: Where the project insights are stored.
+
+    Returns:
+        List of ProjectInsight objects sorted by analyzed_at timestamp.
+    """
```
Comment (ThunderIW) [inline src/project_insights.py]: Good function to get the project sorted

**Review by ThunderIW — APPROVED** [review]
Overall, thank you for making the changes @Weebtrain recommended, and good Job on implementing the last bits of our milestone 1 requirements. PR is approved
